server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -0,0 +1,233 @@
+
+    private static final Logger logger = LogManager.getLogger(LocalHealthMonitor.class);
+
+    public static final Setting<TimeValue> INTERVAL_SETTING = Setting.timeSetting(
+        "health.reporting.local.monitor.interval",
+        TimeValue.timeValueSeconds(30),
+        TimeValue.timeValueSeconds(10),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private final DiskCheck diskCheck;
+
+    private volatile TimeValue monitorInterval;
+    private volatile boolean enabled;
+    // Signals that the health metadata is available, so we can start monitoring.
+    private volatile boolean healthMetadataInitialized;
+    // Ensures that there will no parallel executions of the monitoring process and
+    // simplifies the rescheduling during enabling/disabling or the change of the interval.
+    private final AtomicBoolean inProgress = new AtomicBoolean();
+    // Keeps the latest health state that was successfully reported.
+    private IndividualNodeHealth lastReportedHealth = null;
+
+    public LocalHealthMonitor(Settings settings, ClusterService clusterService, NodeService nodeService, ThreadPool threadPool) {
+        this.threadPool = threadPool;
+        this.monitorInterval = INTERVAL_SETTING.get(settings);
+        this.enabled = HealthNodeTaskExecutor.ENABLED_SETTING.get(settings);
+        this.clusterService = clusterService;
+        this.diskCheck = new DiskCheck(nodeService);
+        clusterService.addListener(this);
+        ClusterSettings clusterSettings = clusterService.getClusterSettings();
+        clusterSettings.addSettingsUpdateConsumer(INTERVAL_SETTING, this::setMonitorInterval);
+        clusterSettings.addSettingsUpdateConsumer(HealthNodeTaskExecutor.ENABLED_SETTING, this::setEnabled);
+    }
+
+    void setMonitorInterval(TimeValue monitorInterval) {
+        this.monitorInterval = monitorInterval;
+        scheduleNowIfEnabled();
+    }
+
+    void setEnabled(boolean enabled) {
+        this.enabled = enabled;
+        scheduleNowIfEnabled();
+    }
+
+    private void scheduleNextRunIfEnabled(TimeValue time) {
+        if (enabled && healthMetadataInitialized) {
+            threadPool.scheduleUnlessShuttingDown(time, ThreadPool.Names.MANAGEMENT, this::monitorHealth);
+        }
+    }
+
+    // Helper method that starts the monitoring without a delay.
+    private void scheduleNowIfEnabled() {
+        scheduleNextRunIfEnabled(new TimeValue(1));
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // Wait until every node in the cluster is upgraded to 8.4.0 or later
+        if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)) {
+            // Wait until the health metadata is available in the cluster state
+            if (healthMetadataInitialized == false) {
```

Comment:
This flag is a little misleading as it waits for a bit more than just the metadata to be initialised (ie. nodes versions). Could we rename it to reflect that?

Also, if the service is enabled,disabled, and enabled again we might end up with 2 scheduled monitoring tracks? (since `setEnabled` doesn't check if we're already running?)

Reply:
About the scheduling two monitoring tasks. That's true, but then one will see that the `inProgress` flag is not false and it will just exit without doing anything or scheduling anything. I thought it was good trade-off compared to locking. What do you think?

Reply:
Yes ! That's a good trade-off 🚀  Can you please document it?

Reply:
Done, let me know if it's clear.

Reply:
Also renamed `healthMetadataInitialized` :)